browser/brave_shell_integration_win.cc

```diff
@@ -0,0 +1,170 @@
+
+namespace {
+
+std::wstring ExtractShortcutNameFromProperties(
+    const ShellUtil::ShortcutProperties& properties) {
+  std::wstring shortcut_name = properties.has_shortcut_name()
+                                   ? properties.shortcut_name
+                                   : InstallUtil::GetShortcutName();
+
+  if (!base::EndsWith(shortcut_name, installer::kLnkExt,
+                      base::CompareCase::INSENSITIVE_ASCII))
+    shortcut_name += installer::kLnkExt;
+
+  return shortcut_name;
+}
+
+// Referenced to FF's implementation.
+// https://github.com/mozilla/gecko-dev/blob/master/browser/components/shell/nsWindowsShellService.cpp
+void PinToTaskbarWin10(const base::FilePath& shortcut_path) {
+  DCHECK(base::win::GetVersion() >= base::win::Version::WIN10_RS5);
+
+  // This enum is likely only used for Windows telemetry, INT_MAX is chosen to
+  // avoid confusion with existing uses.
+  enum PINNEDLISTMODIFYCALLER { PLMC_INT_MAX = INT_MAX };
+
+  // The types below, and the idea of using IPinnedList3::Modify,
+  // are thanks to Gee Law <https://geelaw.blog/entries/msedge-pins/>
+  static constexpr GUID CLSID_TaskbandPin = {
+      0x90aa3a4e,
+      0x1cba,
+      0x4233,
+      {0xb8, 0xbb, 0x53, 0x57, 0x73, 0xd4, 0x84, 0x49}};
+
+  static constexpr GUID IID_IPinnedList3 = {
+      0x0dd79ae2,
+      0xd156,
+      0x45d4,
+      {0x9e, 0xeb, 0x3b, 0x54, 0x97, 0x69, 0xe9, 0x40}};
+
+  struct IPinnedList3Vtbl;
+  struct IPinnedList3 {
+    IPinnedList3Vtbl* vtbl;
+  };
+
+  typedef ULONG STDMETHODCALLTYPE ReleaseFunc(IPinnedList3 * that);
+  typedef HRESULT STDMETHODCALLTYPE ModifyFunc(
+      IPinnedList3 * that, PCIDLIST_ABSOLUTE unpin, PCIDLIST_ABSOLUTE pin,
+      PINNEDLISTMODIFYCALLER caller);
+
+  struct IPinnedList3Vtbl {
+    void* QueryInterface;  // 0
+    void* AddRef;          // 1
+    ReleaseFunc* Release;  // 2
+    void* Other[13];       // 3-15
+    ModifyFunc* Modify;    // 16
+  };
+
+  struct ILFreeDeleter {
+    void operator()(LPITEMIDLIST ptr) {
+      if (ptr) {
+        ILFree(ptr);
+      }
+    }
+  };
+
+  IPinnedList3* pinned_list = nullptr;
+  HRESULT hr = CoCreateInstance(CLSID_TaskbandPin, nullptr,
+                                CLSCTX_INPROC_SERVER, IID_IPinnedList3,
+                                reinterpret_cast<void**>(&pinned_list));
+  if (FAILED(hr) || !pinned_list) {
+    LOG(ERROR) << __func__ << " CoCreateInstance failed.";
+    return;
+  }
+
+  std::unique_ptr<__unaligned ITEMIDLIST, ILFreeDeleter> path(
+      ILCreateFromPathW(shortcut_path.value().c_str()));
+  hr =
+      pinned_list->vtbl->Modify(pinned_list, nullptr, path.get(), PLMC_INT_MAX);
+  pinned_list->vtbl->Release(pinned_list);
+
+  if (FAILED(hr))
+    LOG(ERROR) << __func__ << " Failed to pin to taskbar.";
+}
+
+void PinToTaskbarImpl() {
+  base::FilePath chrome_exe;
+  if (!base::PathService::Get(base::FILE_EXE, &chrome_exe))
+    return;
+
+  ShellUtil::ShortcutProperties properties(ShellUtil::CURRENT_USER);
```

Comment:
Q. If we call `properties.set_pin_to_taskbar(true)`, then ShellUtil::PinShortcut() will be called. https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/util/shell_util.h;l=865

To me, Chromium's implementation looks quite similar to FF's implementation but more modern as they're using ComPtr(RAII obj?). Doesn't Chromium's implementation work and is it why you brought implementation from FF?

Reply:
> Q. If we call `properties.set_pin_to_taskbar(true)`, then ShellUtil::PinShortcut() will be called. https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/util/shell_util.h;l=865

`properties.set_pin_to_taskbar(true)` will not work for Win10 and later version. 
AFAIK, `taskbarpin` is disabled by MS from Win10.
```
bool CanPinShortcutToTaskbar() {
  // "Pin to taskbar" stopped being supported in Windows 10.
  return GetVersion() < Version::WIN10;
}

bool PinShortcutToTaskbar(const FilePath& shortcut) {
  ScopedBlockingCall scoped_blocking_call(FROM_HERE, BlockingType::MAY_BLOCK);
  DCHECK(CanPinShortcutToTaskbar());

  intptr_t result = reinterpret_cast<intptr_t>(ShellExecute(
      nullptr, L"taskbarpin", shortcut.value().c_str(), nullptr, nullptr, 0));
  return result > 32;
}
```

> 
> To me, Chromium's implementation looks quite similar to FF's implementation but more modern as they're using ComPtr(RAII obj?). Doesn't Chromium's implementation work and is it why you brought implementation from FF?

Hmm, it's not yet available for our master. Should we copy and test in advance?

Reply:
Trying to test chromium's method.

Reply:
Oh, there's another implementation in base::win::shortcut.cc , which uses "ShellExecute()" . Maybe this has been stopped working from Win10.

> Hmm, it's not yet available for our master

Aha, What I found was the latest one and we don't have it yet. Thanks.